ml-agents/mlagents/trainers/settings.py

```diff
@@ -0,0 +1,365 @@
+        memory_size: int = 128
+
+    normalize: bool = False
+    hidden_units: int = 128
+    num_layers: int = 2
+    vis_encode_type: EncoderType = EncoderType.SIMPLE
+    memory: Optional[MemorySettings] = None
+
+
+@attr.s(auto_attribs=True)
+class BehavioralCloningSettings:
+    demo_path: str
+    steps: int = 0
+    strength: float = 1.0
+    samples_per_update: int = 0
+    num_epoch: Optional[int] = None
+    batch_size: Optional[int] = None
+
+
+@attr.s(auto_attribs=True)
+class HyperparamSettings:
+    batch_size: int = 1024
+    buffer_size: int = 10240
+    learning_rate: float = 3.0e-4
+    learning_rate_schedule: LearningRateSchedule = LearningRateSchedule.CONSTANT
+
+
+@attr.s(auto_attribs=True)
+class PPOSettings(HyperparamSettings):
+    beta: float = 5.0e-3
+    epsilon: float = 0.2
+    lambd: float = 0.95
+    num_epoch: int = 3
+    learning_rate_schedule: LearningRateSchedule = LearningRateSchedule.LINEAR
+
+
+@attr.s(auto_attribs=True)
+class SACSettings(HyperparamSettings):
+    batch_size: int = 128
+    buffer_size: int = 50000
+    buffer_init_steps: int = 0
+    tau: float = 0.005
+    steps_per_update: float = 1
+    save_replay_buffer: bool = False
+    init_entcoef: float = 1.0
+    reward_signal_steps_per_update: float = attr.ib()
+
+    @reward_signal_steps_per_update.default
+    def _reward_signal_steps_per_update_default(self):
+        return self.steps_per_update
+
+
+@attr.s(auto_attribs=True)
+class RewardSignalSettings:
+    class RewardSignalType(Enum):
+        EXTRINSIC: str = "extrinsic"
+        GAIL: str = "gail"
+        CURIOSITY: str = "curiosity"
+
+    @staticmethod
+    def to_settings(ttype: RewardSignalType) -> type:
+        _mapping = {
+            RewardSignalSettings.RewardSignalType.EXTRINSIC: RewardSignalSettings,
+            RewardSignalSettings.RewardSignalType.GAIL: GAILSettings,
+            RewardSignalSettings.RewardSignalType.CURIOSITY: CuriositySettings,
+        }
+        return _mapping[ttype]
+
+    gamma: float = 0.99
+    strength: float = 1.0
+
+
+@attr.s(auto_attribs=True)
+class GAILSettings(RewardSignalSettings):
+    encoding_size: int = 64
+    learning_rate: float = 3e-4
+    use_actions: bool = False
+    use_vail: bool = False
+    demo_path: str = attr.ib(kw_only=True)
+
+
+@attr.s(auto_attribs=True)
+class CuriositySettings(RewardSignalSettings):
+    encoding_size: int = 128
+    learning_rate: float = 3e-4
+
+
+@attr.s(auto_attribs=True)
+class SelfPlaySettings:
+    save_steps: int = 20000
+    team_change: int = attr.ib()
+
+    @team_change.default
+    def _team_change_default(self):
+        # Assign team_change to about 4x save_steps
+        return self.save_steps * 5
+
+    swap_steps: int = 10000
+    window: int = 10
+    play_against_latest_model_ratio: float = 0.5
+    initial_elo: float = 1200.0
+
+
+@attr.s(auto_attribs=True)
+class TrainerSettings:
+    # Edit these two fields to add new trainers #
+    class TrainerType(Enum):
```

Comment:
IMO this class and the associated static method should live fully outside of `TrainerSettings`.  

So...

```python
class TrainerType(Enum):
  PPO: str = "ppo"
  SAC: str = "sac"

  def to_settings(self) -> type:
    _mapping = {
      self.PPO: PPOSettings,
      self.SAC: SACSettings,
    }
    return _mapping[self]
```

I also used self-references for the members, which might arguably be less clear.  That isn't necessary.

Reply:
Split off the type enums (TrainerType and RewardSIgnalType) as separate classes.